src/third/redux-persist/createPersistor.js

```diff
@@ -57,9 +57,10 @@ export default function createPersistor (store, config) {
         writes.push([key, serializer(substate)])
       }
 
-      writes.forEach(([key, serializedSubstate]) => {
-        storage.setItem(createStorageKey(key), serializedSubstate).catch(warnIfSetError(key))
-      })
+      // Warning: not guaranteed to be done in a transaction.
+      storage.multiSet(
+        writes.map(([key, serializedSubstate]) => [createStorageKey(key), serializedSubstate])
+      ).catch(warnIfSetError(updatedSubstates));
```

Comment:
This argument to `warnIfSetError` is always an empty array at this point, right?

The code around this changes later in the branch, but a thing that'd be good to change about the final code would actually already help here. In the code at the tip, we hang on to the whole `updatedSubstates` including all the substate data:
```js
    for (const [key, substate] of updatedSubstates) {
      writes.push([key, serializer(substate)])
      await new Promise(r => setTimeout(r, 0));
    }
```
even as we generate the serialized strings. For something like `state.users`, that can be quite a big string and also a big `User[]` array, which this code might be holding the last reference to if there've been subsequent updates. It'd be good to discard each subtree as soon as we've serialized it out.

The code as of this commit does that, by mutating `updatedSubstates`. That's probably the most natural way to keep doing it, even later in the series.

Then in order to keep around the list of keys for the sake of this error logging, the serialization loop could push each key onto a new array that's just the keys. Like this:
```js
       while (updatedSubstates.length > 0) {
         const [key, substate] = updatedSubstates.shift();
         writes.push([createStorageKey(key), serializer(substate)]);
         keys.push(key);
         await new Promise(r => setTimeout(r, 0));
       }
```

Reply:
> For something like `state.users`, that can be quite a big string and also a big `User[]` array, which this code might be holding the last reference to if there've been subsequent updates. It'd be good to discard each subtree as soon as we've serialized it out.

Hmm, I guess actually we hang on to each subtree in any case, in the form of `lastState`, in order to be able to check later if it's changed. So there isn't really a benefit to be had from discarding it from `updatedSubstates`.

This memory-consumption issue gets solved by converting the large data structures to Immutable -- if we had an `Immutable.List<User>` instead of a `User[]` there, then it'd share most of its memory with later versions of itself. So that will make it definitely not a concern in the future.

If it were really essential to solve it in a context where we did still have large arrays that didn't get to share memory like that, we could do it by having `lastState` keep around only weak references (using [JS's `WeakRef`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef), probably.) That'd ensure that `lastState` never caused a piece of state to stick around when it otherwise wouldn't, and it'd still give us all the information we need because the only thing we really care about in an old state subtree is whether it's the same value that is the new state subtree. But I don't think the complication is worth it.

As of this commit, you could do the `keys.push(key)` thing, or could also just have the log message not mention the keys -- that'd be fine as an intermediate state in the branch.